### PR TITLE
Feature/refactor virtualization

### PR DESCRIPTION
**Pull request by talekhinezh**

### CODE REVIEW

**Review by siy — APPROVED** [review]
LGTM

File: radixdlt-engine/src/main/java/com/radixdlt/application/system/scrypt/SystemConstraintScrypt.java [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+				},
+				buf -> {
+					var addr = REFieldSerialization.deserializeREAddr(buf, claimableAddrTypes);
+					return new UnclaimedREAddr(addr);
+				},
+				(a, buf) -> REFieldSerialization.serializeREAddr(buf, (REAddr) a)
+			));
+
+		os.procedure(new DownProcedure<>(
+			REAddrClaimStart.class, UnclaimedREAddr.class,
```
Comment (iamyulong) [inline radixdlt-engine/src/main/java/com/radixdlt/application/system/scrypt/SystemConstraintScrypt.java:156]: Could you help me understand how we ensure each REAddress is claimed once only?

File: radixdlt-engine/src/main/java/com/radixdlt/constraintmachine/CallData.java [hunk 2]
```diff
@@ -44,6 +46,23 @@ public UInt256 getUInt256(int offset) throws CallDataAccessException {
...
+		if (offset < 0 || offset > data.length) {
+			throw new CallDataAccessException(data.length, offset, 1);
+		}
+
+		var length = Byte.toUnsignedInt(data[offset]);
+		if (length < 1 || length > 32) {
+			throw new ProcedureException("Invalid length");
+		}
+
+		if (length + offset > data.length) {
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radixdlt-engine/src/main/java/com/radixdlt/constraintmachine/CallData.java:59]: good catch

File: radixdlt-engine/src/main/java/com/radixdlt/constraintmachine/REOp.java [hunk 3]
```diff
@@ -23,6 +23,7 @@
 	UP(true),
 	READ(false),
 	DOWN(true),
+	ARG(false),
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radixdlt-engine/src/main/java/com/radixdlt/constraintmachine/REOp.java:26]: good catch

File: radixdlt-engine/src/main/java/com/radixdlt/atomos/SubstateDefinition.java [hunk 4]
```diff
@@ -46,21 +47,28 @@ public SubstateDefinition(
...
 	}
 
 	public SubstateDefinition(
 		Class<T> substateClass,
 		byte typeByte,
 		SubstateDeserializer<T> deserializer,
 		SubstateSerializer<T> serializer,
-		Predicate<T> virtualized
+		SubstateDeserializer<T> virtualDeserializer,
+		VirtualSubstateSerializer virtualSubstateSerializer
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radixdlt-engine/src/main/java/com/radixdlt/atomos/SubstateDefinition.java:64]: We really want virtualization to be defined by the key that is virtualized. The bugs that have been occurring is an incorrect `Predicate<T>` which allows a given key to have multiple virtualized substates. The fix here is to define `Predicate<T>` by explicitly defining `Function<Key, T>` for virtualized state.

File: radixdlt-engine/src/main/java/com/radixdlt/application/system/scrypt/SystemConstraintScrypt.java (hunk 1 above)
Comment (talekhinezh) [reply radixdlt-engine/src/main/java/com/radixdlt/application/system/scrypt/SystemConstraintScrypt.java:156]: Each REAddr has one and only one virtual `UnclaimedREAddr` substate. When claimed that substate is shutdown and thus can never be claimed again.

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/store/berkeley/BerkeleyLedgerEntryStore.java [hunk 5]
```diff
@@ -367,24 +367,24 @@ private void open() {
...
+								// 3-6: Epoch
+								prefixIndexSize = 3 + Long.BYTES;
+							} else if (substateTypeId == SubstateTypeId.VALIDATOR_RAKE_COPY.id()) {
 								// 0: Type Byte
 								// 1: Reserved Byte
-								// 2-5: Epoch
-								prefixIndexSize = 2 + Long.BYTES;
+								// 2: Optional flag
+								// 3-6: Epoch
+								prefixIndexSize = 3 + Long.BYTES;
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/store/berkeley/BerkeleyLedgerEntryStore.java:387]: I think cleaner to be explicit

**Review by iamyulong — APPROVED** [review]

**Pull request merged**